### ISSUE TITLE
[question] How to mock functions/headers from conan pkgs in cmake

### ISSUE THREAD

**Issue by daltonv** [conan-io/conan#15008]
### What is your question?

I've been trying to integrate the [cmock mocking framework](https://github.com/ThrowTheSwitch/CMock) into my build system that uses CMake and Conan packages. 

CMock is simply a ruby script that you point at a header file and it create a mocked version of that header file with a corresponding c file that can be used for unit tests. 

Where I run into difficulties is when I want to mock header files that originate from a Conan pkg. To call CMock properly I must give an absolute path to the header file to mock, and I don't want to have to know the location on disk of the Conan package I'm using. 

I've tried to write a CMake script that would recursively grab a list of include directories for a given target and its dependent libs (eventually conan ones). Something like this:
```cmake
function(get_recursive_interface_includes target includes)
    get_target_property(interface_libraries ${target} INTERFACE_LINK_LIBRARIES)
    foreach(interface_library ${interface_libraries})
        get_target_property(interface_includes ${interface_library} INTERFACE_INCLUDE_DIRECTORIES)
        list(APPEND includes ${interface_includes})
        get_recursive_interface_includes(${interface_library} includes)
    endforeach()
endfunction()
```

Unfortunately this ultimately fails because the conan packages generate targets that use generator expressions like `$<$<CONFIG:MinSizeRel>:${mspm0sdk_mspm0sdk_driverlib_OBJECTS_MINSIZEREL}`, so these cannot be used in config time for cmake.

So to sum up I want a CMake script that you give as an input a header path relative to your include paths for a given target, and you get as an output the absolute path of that header. With my current approach Conan gets in the way with generator expressions.

Has anyone else ran into a problem like this, or have any ideas on how to solve it?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @daltonv 

Thanks for your question.
Have you considered using the ``generate()`` method to generate a .cmake file with the information you need? It would be relatively straightforward, something like:

```python
def generate(self):
      cmakefile = ""
      for req, dep in self.dependencies.host.items():
           includedirs = dep.cpp_info.includedirs
           for includedir in includedirs:
                 cmakefile += f"....{includedir}"
     save(self, "myheaders.cmake", cmakefile)
```
then you can include the ``myheaders.cmake`` file and you will have the processed information there.

**Comment by daltonv** [comment]
Hi @memsharded,

 Not a bad idea. I could have my mock script just always search those headers in addition to headers normally accessible from a target. I think this should unblock me for now, but it still feels like it could be more generic. 

Ideally I'd like the solution to work without having anything special handling for Conan. It feels like I should be able to retrieve this info from any CMake package, but I'm also relatively new to CMake though, so what I'm trying might just be unworkable in CMake by itself.

**Comment by memsharded** [comment]
> Ideally I'd like the solution to work without having anything special handling for Conan. It feels like I should be able to retrieve this info from any CMake package, but I'm also relatively new to CMake though, so what I'm trying might just be unworkable in CMake by itself.

This can indeed be very tricky for multi-configuration setups, because of the generator expressions. It is true that in many cases, it shouldn't be necessary to have a generator expression for the headers, but the reality is that there are cases that headers are different based on build-types, due to generated code. And the same approach should be able to work in case we wanted to gather the libraries or library directories, which would definitely be different for different build-types, and consequently require generator expressions (there are mocking frameworks that are able to patch binaries on the fly too).

There are also other factors that could have issues in the cmake approach, for example managing the visibility of transitive headers, that wouldn't be necessary to mock. Conan has the information in the ``requirement traits``:

```python
def generate(self):
     cmakefile = ""
     for req, dep in self.dependencies.host.items():
          if req.headers:  # the requirement trait tells me the consumer has visibility over these headers
              includedirs = dep.cpp_info.includedirs
              for includedir in includedirs:
                 cmakefile += f"....{includedir}"
     save(self, "myheaders.cmake", cmakefile)
```

So even if it is not a pure CMake solution, it is probably quite a pragmatic one, easy to understand and maintain.

**Comment by daltonv** [comment]
Gotcha thanks for the detailed explanation. I'll live with the custom generator solution.
